models/user/userModel.js

```diff
@@ -0,0 +1,66 @@
+import mongoose, { Schema } from 'mongoose';
+import { addressSchema, orderItemSchema } from '../schemas';
+import { accountStatuses, roles } from '../../constants';
+import { phoneNumberPattern } from '../../helpers';
+
+
+const ObjectId = Schema.Types.ObjectId;
+
+const userSchema = new Schema(
+  {
+    fullName: {
+      type: String,
+      required: [true, 'Full name is required'],
+      minlength: [3, 'Full name must be at least 3 characters long'],
+    },
+    password: {
+      type: String,
+      required: [true, 'Password is required'],
+      minlength: [8, 'Password must be at least 8 characters long'],
+    },
+    email: {
+      type: String,
+      required: [true, 'Email is required'],
+      unique: true,
+      validate: {
+        validator: (value) => /\S+@\S+\.\S+/.test(value),
+        message: 'Invalid email format',
+      },
+    },
+    avatar: String,
+    address: addressSchema,
+    phoneNumber: {
+      type: String,
+      validate: {
+        validator: (value) => phoneNumberPattern.test(value),
+        message: 'Invalid phone number format',
+      },
+    },
+    favoriteDishes: [{ type: ObjectId, ref: 'dish' }],
+    favoriteChefs: [{ type: ObjectId, ref: 'chef' }],
+    cart: [orderItemSchema],
+    roles: {
+      type: [String],
```

Comment:
multiple roles, for what purpose?

Reply:
Yep, one user can have multiple roles. By default, when a user registers, their role is USER, but they can also go through another registration process with the same address and password to become a CHEF or a COURIER. As a result, all the roles registered by the user are planned to be stored here, in the role array

Reply:
complicated logic that needs to be handled accordingly, but okay

Reply:
We decided to use this role array for the authorization process. It was changed a bit:
This array will have, for example, the following form:
```
roles: [
{ 
name: 'chef',
id: "507f1f77bcf86cd799439011"
},
{
name: "courier",
id: null
},
{name: "admin",
id: null
}
]
```

During token verification we receive userId, and a role parameter for a specific endpoint, for example:
```
router.get('/protectedRoute', verifyToken({ role: 'chef' }), async (req, res) => {
 *   // Your protected route logic here
 * });
```

As a result of authorization, we iterate through the roles array, searching for the assigned role. If the id for that role exists, we pass it; otherwise, we issue an 'access denied'

Reply:
so, we write _id from User model into JWT token, when user authenticates:
1. we decode token and get that _id
2. we search user from User collection by _id
3. after user is found we check roles array for needed role
4. if role is there
  4.1. we get id from it
  4.2. we search for a Chef, for example, with that id
  4.3. if Chef document is in our database and it's an active verified user, we let it in
5. if there's no needed role or something from step 4 does not fir - we deny authentication

that's correct explanation?

I will point my concerns here:
1.  how will we manage tokens if user was authenticated first and then decided to switch to his chef account?
2. if user has chef account, but logged in into user account and puts chef endpoint into address line, what will happen then? by this logic, we will check his roles, find a chef there and let him in? but he was logged in as a user, not chef

Reply:
I didn't add step 4.2 in the auth middleware:
`4.2. We search for a Chef, for example, with that ID.`
Should it be done in auth middleware or inside each endpoint logic?

We don't plan to change the token if it's the same user but another role. The token will have a user's id as a payload. The reason for using this approach is to allow the user to change a role by changing the route in the browser. In this case, when he changes a route, he will work with different endpoints with predefined role parameters accordingly

for user it's
`router.get('/protectedRoute', verifyToken({ role: 'user' }), async (req, res) => {`
for chef it's
`router.get('/protectedRoute', verifyToken({ role: 'chef' }), async (req, res) => {`

Reply:
UPD:
Some routes can be used by several roles, so I changed the auth middleware to allow passing an array of required roles instead of just one role.

Example usage: `router.get('/api/orders/by-chef/:chefId', verifyToken(['chef', 'admin']), async (req, res) => {}`

In this case the middleware returns `{ userId }` and roles: `{ chef: chefId, admin: adminId }`

Reply:
it's complicated logic which in some way breaks the aim of endpoint protection if user can operate chef's endpoints as well
if you are sure this is desired - okay

Reply:
The main purpose of this approach is to be able to change roles, for example from user to chef, without having the re-login. Just by changing the route in the browser and adding /chef/..... So this would immediately change the role, and there would be an interface for the chef role.

I would really appreciate it if you could tell me from your experience, in this case what the disadvantages of this authentication method are, maybe there is some way to strengthen the verification somehow